src/test/groovy/com/netgrif/application/engine/auth/LoginAttemptsTest.groovy

```diff
@@ -0,0 +1,200 @@
+                .with(SecurityMockMvcRequestPostProcessors.authentication(getAuth(USER_EMAIL, USER_PASSWORD, "fakeIp2"))))
+                .andExpect(MockMvcResultMatchers.status().isOk())
+                .andReturn()
+        assert result
+    }
+
+    @Test
+    void loginBad() {
+
+        def result = mvc.perform(MockMvcRequestBuilders.get(LOGIN_URL)
+                .accept(MediaTypes.HAL_JSON_VALUE)
+                .contentType(MediaType.APPLICATION_JSON_UTF8_VALUE)
+                .with(SecurityMockMvcRequestPostProcessors.csrf().asHeader())
+                .with(SecurityMockMvcRequestPostProcessors.authentication(getAuth(USER_EMAIL, USER_BAD_PASSWORD, "fakeIp"))))
+                .andExpect(MockMvcResultMatchers.status().isUnauthorized())
+                .andReturn()
+        assert result
+        result = mvc.perform(MockMvcRequestBuilders.get(LOGIN_URL)
+                .accept(MediaTypes.HAL_JSON_VALUE)
+                .contentType(MediaType.APPLICATION_JSON_UTF8_VALUE)
+                .with(SecurityMockMvcRequestPostProcessors.csrf().asHeader())
+                .with(SecurityMockMvcRequestPostProcessors.authentication(getAuth(USER_EMAIL, USER_BAD_PASSWORD, "fakeIp"))))
+                .andExpect(MockMvcResultMatchers.status().isUnauthorized())
+                .andReturn()
+        assert result
+        result = mvc.perform(MockMvcRequestBuilders.get(LOGIN_URL)
+                .accept(MediaTypes.HAL_JSON_VALUE)
+                .contentType(MediaType.APPLICATION_JSON_UTF8_VALUE)
+                .with(SecurityMockMvcRequestPostProcessors.csrf().asHeader())
+                .with(SecurityMockMvcRequestPostProcessors.authentication(getAuth(USER_EMAIL, USER_BAD_PASSWORD, "fakeIp"))))
+                .andExpect(MockMvcResultMatchers.status().isUnauthorized())
+                .andReturn()
+        assert result
+        result = mvc.perform(MockMvcRequestBuilders.get(LOGIN_URL)
+                .accept(MediaTypes.HAL_JSON_VALUE)
+                .contentType(MediaType.APPLICATION_JSON_UTF8_VALUE)
+                .with(SecurityMockMvcRequestPostProcessors.csrf().asHeader())
+                .with(SecurityMockMvcRequestPostProcessors.authentication(getAuth(USER_EMAIL, USER_BAD_PASSWORD, "fakeIp"))))
+                .andExpect(MockMvcResultMatchers.status().isUnauthorized())
+                .andReturn()
+        assert result
+        //BLOCK user
+        result = mvc.perform(MockMvcRequestBuilders.get(LOGIN_URL)
+                .accept(MediaTypes.HAL_JSON_VALUE)
+                .contentType(MediaType.APPLICATION_JSON_UTF8_VALUE)
+                .with(SecurityMockMvcRequestPostProcessors.csrf().asHeader())
+                .with(SecurityMockMvcRequestPostProcessors.authentication(getAuth(USER_EMAIL, USER_PASSWORD, "fakeIp"))))
+                .andExpect(MockMvcResultMatchers.status().isUnauthorized())
+                .andReturn()
+        assert result
+        sleep(5000) //Wait
```

Comment:
the duration of the sleep must presumably be greater than some value specified in some configuration. It would be better to reference this value in the test, so that, should it change, the test would not fail.

Reply:
Done